sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala

```diff
@@ -191,6 +216,216 @@ class RapidsHostMemoryStore(
+        }
+      }
+    }
+  }
+
+  /**
+   * A `RapidsHostColumnarBatch` is the spill store holder of ColumnarBatch backed by
+   * HostColumnVector.
+   *
+   * This class owns the host batch and will close it when `close` is called.
+   *
+   * @param id the `RapidsBufferId` this batch is associated with
+   * @param batch the host ColumnarBatch we are managing
+   * @param spillPriority a starting spill priority
+   */
+  class RapidsHostColumnarBatch(
+      id: RapidsBufferId,
+      hostCb: ColumnarBatch,
+      spillPriority: Long)
+      extends RapidsBufferBase(
+        id,
+        null,
+        spillPriority) {
+
+    override val storageTier: StorageTier = StorageTier.HOST
+
+    override val needsSerialization: Boolean = true
+
+    // This is the current size in batch form. It is to be used while this
+    // batch hasn't migrated to another store.
+    private val hostSizeInByes: Long = RapidsHostColumnVector.getTotalHostMemoryUsed(hostCb)
+
+    // By default all columns are NOT spillable since we are not the only owners of
+    // the columns (the caller is holding onto a ColumnarBatch that will be closed
+    // after instantiation, triggering onClosed callbacks)
+    // This hash set contains the columns that are currently spillable.
+    private val columnSpillability = new ConcurrentHashMap[HostColumnVector, Boolean]()
+
+    private val numDistinctColumns = RapidsHostColumnVector.extractBases(hostCb).distinct.size
+
+    // we register our event callbacks as the very first action to deal with
+    // spillability
+    registerOnCloseEventHandler()
+
+    /** Release the underlying resources for this buffer. */
+    override protected def releaseResources(): Unit = {
+      hostCb.close()
+    }
+
+    override def meta: TableMeta = {
+      null
+    }
+
+    override def getMemoryUsedBytes: Long = hostSizeInByes
+
+    /**
+     * Mark a column as spillable
+     *
+     * @param column the ColumnVector to mark as spillable
+     */
+    def onColumnSpillable(column: HostColumnVector): Unit = {
+      columnSpillability.put(column, true)
+      updateSpillability()
+    }
+
+    /**
+     * Update the spillability state of this RapidsHostColumnarBatch. This is invoked from
+     * two places:
+     *
+     * - from the onColumnSpillable callback, which is invoked from a
+     * HostColumnVector.EventHandler.onClosed callback.
+     *
+     * - after adding a batch to the store to mark the batch as spillable if
+     * all columns are spillable.
+     */
+    override def updateSpillability(): Unit = {
+      doSetSpillable(this, columnSpillability.size == numDistinctColumns)
+    }
+
+    override def getColumnarBatch(sparkTypes: Array[DataType]): ColumnarBatch = {
+      throw new UnsupportedOperationException(
+        "RapidsHostColumnarBatch does not support getColumnarBatch")
+    }
+
+    override def getHostColumnarBatch(sparkTypes: Array[DataType]): ColumnarBatch = {
+      columnSpillability.clear()
```

Comment:
How is this going to interact with another thread that's trying to call onColumnSpillable at the same time?  Could we sometimes end up in a state where we built the host batch but still think this is spillable?  I think the answer is yes, but maybe we don't care because of the refcounts?  In light of that, seems prudent to first get the refcounts in place before marking this as spillable, otherwise I think another thread could end up calling `releaseResources` and close hostCb on this (now spillable) buffer before we get around to locking down hostCb.

Reply:
`getHostColumnarBatch`, like `getColumnarBatch` must be called after acquiring the `RapidsBuffer` via `catalog.acquireBuffer`. It will already have dealt with the `RapidsBuffer` reference count, so we are guaranteed that the batch produced isn't pointing to a closed batch.

And yes I see the race you mention. I'll think about that more, to see if we need to lock things further.

Reply:
One thing we don't do is mark something as not spillable if it was acquired. It is only non-spillable when the batch or buffer is obtained.

Perhaps we should add a second level of unspillable. While a buffer has RapidsBuffer.refCount > 1, we could mark the buffer as non spillable. When the RapidsBuffer.refCount goes back to 1, we could return to the cuDF-driven spillability checks.

Reply:
BTW this is an issue with device buffers and batches as well

Reply:
> While a buffer has RapidsBuffer.refCount > 1, we could mark the buffer as non spillable.

This seems desirable in general, because it's not worth spilling something if we cannot the free memory associated with it after the spill copy.  Not something to tackle in this PR.

Reply:
Thanks @jlowe. I filed this issue to track https://github.com/NVIDIA/spark-rapids/issues/9120